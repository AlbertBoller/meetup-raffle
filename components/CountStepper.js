import React from 'react';
import PropTypes from 'prop-types';
import Stepper from 'react-stepper-primitive';

const CountStepper = ({ field, form: { setFieldValue } }) => (
  <Stepper
    min={1}
    max={9}
    defaultValue={field.value}
    onChange={value => setFieldValue(field.name, value)}
    render={({ getInputProps, getIncrementProps, getDecrementProps }) => (
      <span className="mh3-ns">
        <button
          type="button"
          className="bn bg-transparent w2 pointer"
          {...getDecrementProps()}
        >
          <img src="/static/minus.svg" alt="minus" />
        </button>
        <input className="tc w2 pv1" {...field} {...getInputProps()} />
        <button
          type="button"
          className="bn bg-transparent w2 pointer"
          {...getIncrementProps()}
        >
          <img src="/static/plus.svg" alt="plus" />
        </button>
      </span>
    )}
  />
);

CountStepper.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  form: PropTypes.shape({
    setFieldValue: PropTypes.func.isRequired,
  }).isRequired,
};

export default CountStepper;
