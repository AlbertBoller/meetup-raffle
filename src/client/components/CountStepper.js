import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Stepper from 'react-stepper-primitive';

const CountStepper = ({
  inputId,
  labelText,
  field,
  form: { setFieldValue },
}) => (
  <Stepper
    enableReinitialize
    min={1}
    max={9}
    defaultValue={parseInt(field.value, 10)}
    onChange={value => setFieldValue(field.name, value)}
    render={({ getInputProps, getIncrementProps, getDecrementProps }) => (
      <Fragment>
        {labelText && (
          <div className="mb3">
            <label className="f4 f3-ns lh-copy dark-blue" htmlFor={inputId}>
              {labelText}
            </label>
          </div>
        )}
        <span className="flex flex-row">
          <button
            aria-label="decrement"
            type="button"
            className="bn bg-transparent near-black h3 w3 pointer"
            data-testid="decrement-button"
            {...getDecrementProps()}
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              fill="currentColor"
              viewBox="0 0 40 40"
              preserveAspectRatio="xMidYMid meet"
            >
              <g>
                <path d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z" />
              </g>
            </svg>
          </button>
          <input
            id={inputId}
            className="tc near-black w3 f3 pv1"
            {...field}
            {...getInputProps()}
          />
          <button
            aria-label="increment"
            type="button"
            className="bn bg-transparent near-black h3 w3 pointer"
            data-testid="increment-button"
            {...getIncrementProps()}
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              fill="currentColor"
              viewBox="0 0 40 40"
              preserveAspectRatio="xMidYMid meet"
            >
              <g>
                <path d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z" />
              </g>
            </svg>
          </button>
        </span>
      </Fragment>
    )}
  />
);

CountStepper.propTypes = {
  inputId: PropTypes.string.isRequired,
  labelText: PropTypes.string,
  field: PropTypes.shape({
    value: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  form: PropTypes.shape({
    setFieldValue: PropTypes.func.isRequired,
  }).isRequired,
};

CountStepper.defaultProps = {
  labelText: '',
};

CountStepper.displayName = 'CountStepper';

export default CountStepper;
