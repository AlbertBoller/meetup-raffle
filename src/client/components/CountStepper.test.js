import React from 'react';
import { render, Simulate } from 'react-testing-library';
import CountStepper from './CountStepper';

const countStepperProps = {
  inputId: 'foo',
  labelText: 'Count:',
  field: {
    name: 'foo',
    value: 5,
    onChange: jest.fn(),
  },
  form: {
    setFieldValue: jest.fn(),
  },
};

describe('CountStepper', () => {
  const { container, getByLabelText, getByTestId } = render(
    <CountStepper {...countStepperProps} />,
  );

  it('renders', () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it('increments', () => {
    const initialValue = parseInt(
      getByLabelText(countStepperProps.labelText).value,
      10,
    );
    Simulate.click(getByTestId('increment-button'));
    expect(
      parseInt(getByLabelText(countStepperProps.labelText).value, 10),
    ).toBe(initialValue + 1);
  });

  it('decrements', () => {
    const initialValue = parseInt(
      getByLabelText(countStepperProps.labelText).value,
      10,
    );
    Simulate.click(getByTestId('decrement-button'));
    expect(
      parseInt(getByLabelText(countStepperProps.labelText).value, 10),
    ).toBe(initialValue - 1);
  });
});
