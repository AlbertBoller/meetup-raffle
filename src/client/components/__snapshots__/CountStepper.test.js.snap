// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountStepper renders 1`] = `
<div
  class="mb3"
>
  <label
    class="f4 f3-ns lh-copy dark-blue"
    for="foo"
  >
    Count:
  </label>
</div>
`;
